va/va_enc_hevc.h

```diff
@@ -45,6 +45,200 @@ extern "C" {
+        uint32_t deblocking_filter_disable  : 2;
+        /* Reserved for future use. */
+        uint32_t reserved                   : 2;
+    } bits;
+    uint32_t value;
+} VAConfigAttribValEncHEVCFeatures;
+
+/** Attribute value for VAConfigAttribEncHEVCBlockSizes.
+ *
+ * This attribute describes the supported coding tree and transform block
+ * sizes of an HEVC/H.265 encoder configuration
+ */
+typedef union VAConfigAttribValEncHEVCBlockSizes {
+    struct {
+        /** Largest supported size of coding tree blocks.
+         *
+         * CtbLog2SizeY must not be larger than this.
+         */
+        uint32_t log2_max_coding_tree_block_size_minus3    : 2;
+        /** Smallest supported size of coding tree blocks.
+         *
+         * CtbLog2SizeY must not be smaller than this.
+         *
+         * This may be the same as the maximum size, indicating that only
+         * one CTB size is supported.
+         */
+        uint32_t log2_min_coding_tree_block_size_minus3    : 2;
+
+        /** Smallest supported size of luma coding blocks.
+         *
+         * MinCbLog2SizeY must not be smaller than this.
+         */
+        uint32_t log2_min_luma_coding_block_size_minus3    : 2;
+
+        /** Largest supported size of luma transform blocks.
+         *
+         * MaxTbLog2SizeY must not be larger than this.
+         */
+        uint32_t log2_max_luma_transform_block_size_minus2 : 2;
+        /** Smallest supported size of luma transform blocks.
+         *
+         * MinTbLog2SizeY must not be smaller than this.
+         */
+        uint32_t log2_min_luma_transform_block_size_minus2 : 2;
+
+        /** Largest supported transform hierarchy depth in inter
+         *  coding units.
+         *
+         * max_transform_hierarchy_depth_inter must not be larger
+         * than this.
+         */
+        uint32_t max_max_transform_hierarchy_depth_inter   : 2;
+        /** Smallest supported transform hierarchy depth in inter
+         *  coding units.
+         *
+         * max_transform_hierarchy_depth_inter must not be smaller
+         * than this.
+         */
+        uint32_t min_max_transform_hierarchy_depth_inter   : 2;
+
+        /** Largest supported transform hierarchy depth in intra
+         *  coding units.
+         *
+         * max_transform_hierarchy_depth_intra must not be larger
+         * than this.
+         */
+        uint32_t max_max_transform_hierarchy_depth_intra   : 2;
+        /** Smallest supported transform hierarchy depth in intra
+         *  coding units.
+         *
+         * max_transform_hierarchy_depth_intra must not be smaller
+         * than this.
+         */
+        uint32_t min_max_transform_hierarchy_depth_intra   : 2;
+
+        /** Size of the smallest supported coding unit which can
+         *  carry a QP delta.
+         *
+         * Log2MinCuQpDeltaSize must not be smaller than this.
+         */
+        uint32_t log2_min_cu_qp_delta_size_minus3          : 2;
```

Comment:
From HEVC SPEC,  Log2MinCuQpDeltaSize = CtbLog2SizeY − diff_cu_qp_delta_depth,  if diff_cu_qp_delta_depth can be assigned to 0 or 3 for different LCU,  how to set log2_min_cu_qp_delta_size_minus3?

Reply:
@Tianhaol  The intent here is that the limit of a single hardware implementation is going to be the level of the smallest coding unit for which it can usefully vary qp.  If that is, for example, at the 16x16 block size then we set `log2_min_cu_qp_delta_size_minus3 = 1`, and that then implies that the max value of `diff_cu_qp_delta_depth` would be 2 for 64x64 CTBs or 1 for 32x32 CTBs.

Does that fit with how you use it?  If you have hardware which has significantly different implementations at different CTU sizes and therefore doesn't match that setup then we may need to come up with something else.

Reply:
I have no such concern with current HW. but , if some HW support both 64x64 CTBs and 32x32CTBs, and with different CTBs, there are different values , for 64x64CTBs, diff_cu_qp_delta_depth = 3, log2_min_cu_qp_delta_size_minus3 = 0,
for 32x32CTBs , log2_min_cu_qp_delta_size_minus3 = 2.  of course, we are discussing potential risk. is it extendable ?

Reply:
If diff_cu_qp_delta_depth only can be assigned to two values, 3 or 0, not in the range of 0 to 3, how to set this field log2_min_cu_qp_delta_size_minus3?

Reply:
@fhvwy  hi Mark, another question: if log2_min_cu_qp_delta_size_minus3 = 0, it will mean diff_cu_qp_delta_depth  = 3, does application could set a value < 3? if HW just support one setting, such like diff_cu_qp_delta_depth  must equal to 3. how to report it?

Reply:
@Tianhaol @XinfengZhang  I think you're both talking about the same case, which would be for hardware with 64x64 CTBs where the inclusion of `delta_qp()` is either once at the beginning of each CTU or it has to be included in every CU.

I was not expecting a case like that, and the range-based definition I've given does not support it.  Given that it's not going to work as a range then I think we need some sort of mask setup for whether the depth is allowed to be the given value for each CTB size.

So, perhaps a new config attribute `VAConfigAttribValEncHEVCQPDeltaDepthSizes` defining that mask?  If CTB size NxN can support a QP delta depth of D then we set bit `(log2(N) - 3) * 4 + D` to 1 (and those bits are zero otherwise).  In your example case we would set the value `0b1001_0000_0000_0000 = 0x9000` to indicate that two allowed possibilities are for 64x64 CTBs with a depth of 0 or 3.  Would that make sense?

Reply:
@fhvwy how about to add a "log2_max_cu_qp_delta_size_minus3"  which could be used with "log2_min_cu_qp_delta_size_minus3" to describe specific range, such as 3-3, 2-3, 2-2, 0-0 etc.

Reply:
@XinfengZhang  Unfortunately that doesn't support the case of 0 or 3 but not anything in between.

I did a bit more investigation into hardware support here:

Intel Gen9 (Skylake): https://01.org/sites/default/files/documentation/intel-gfx-prm-osrc-skl-vol10-hevc.pdf
Page 122 has the requirement that `diff_cu_qp_delta_depth = 0`.  Since this was the first VAAPI H.265 encode implementation, this value is what ended up hard-coded by users (and which we are now seeking to remove) - e.g. http://git.videolan.org/?p=ffmpeg.git;a=blob;f=libavcodec/vaapi_encode_h265.c;h=3e0af17ea543e9d6f8095c1e28f4cbf47158be1f;hb=HEAD#l546 .

Intel Gen9.5 (Kaby Lake, Coffee Lake, Comet Lake): https://01.org/sites/default/files/documentation/intel-gfx-prm-osrc-kbl-vol10-hevc.pdf
Page 215 has the requirement that `diff_cu_qp_delta_depth = 0`, matching the Gen9 behaviour.

Intel Gen11 (Ice Lake): https://01.org/sites/default/files/documentation/intel-gfx-prm-osrc-icllp-vol11-mediaengines_0.pdf
Page 109 offers that `diff_cu_qp_delta_depth` must be equal to either `0` or `(lcu_size - min_cu_size)`.  `lcu_size` means `CtbSizeY` and `min_cu_size` means `MinCbSizeY`, so this is exactly the case we described above - it will either have a single `delta_qp()` at the start of the CTU (with `diff_cu_qp_delta_depth = 0`) or have `delta_qp()` in every CU (with `diff_cu_qp_delta_depth = CtbSizeY - MinCbSizeY`), but no case in between.  This is the problematic case which means a range won't work, leading to the bit mask suggestion in my previous comment.

I can't find any documentation for Intel Gen12 (Tiger Lake / others) or for any of the AMD encode blocks (VCE/VCN).

Reply:
thanks @fhvwy , yes, Gen12 should be similiar with Gen11 about this setting. I suggest to remove this field from this structure in this PR, and discuss it another new PR. or. rename it to represent cu_qp_delta_enabled_flag
for this case, from media_driver side, we want to suggest diff_cu_qp_delta_depth as 3 for better quality (maybe it will waste some bits for some bitstream, because of additional bits for cu_qp_delta_abs , cu_qp_delta_sign_flag?) .

Reply:
@XinfengZhang  Ok, that's fair.  I think removing the field is the cleanest option for now, so I've done that and stated against the `cu_qp_delta` feature that `diff_cu_qp_delta_depth` must be set to zero in this version (if it gets revisited later with a new attribute then that comment will need to be updated to point to the new attribute).

Reply:
thanks @fhvwy , about this comment change of cu_qp_delta [https://github.com/intel/libva/pull/385/files#diff-6fddfe52dd51a4b8baecfd42b1895e34f8d41f4e7ba69c4298f3802490b81292R124 ] , it means that application must set  diff_cu_qp_delta_depth to be 0 when cu_qp_delta = VA_FEATURE_SUPPORTED or VA_FEATURE_REQUIRED, right?  TBH, it was not my intension, because it cause 2 potential problems: 
1. backward compatible issue in future, even we add one new attribute to represent diff_cu_qp_delta_depth in future, application should always set it to 0.    
2. it is workable for gen9 and gen95, but suppose not workable for VDEnc path of gen11 and gen12
from this point of view, keeping log2_min_cu_qp_delta_size_minus3   and adding log2_max_cu_qp_delta_size_minus3 is better. 
or, keeping log2_min_cu_qp_delta_size_minus3, and change the comments of cu_qp_delta:
/* Allows setting cu_qp_delta_enabled_flag in the PPS.  if it is VA_FEATURE_SUPPORTED diff_cu_qp_delta_depth  range is [ 0 ~ log2 CtbSizeY - 3 - log2_min_cu_qp_delta_size_minus3] , if it is VA_FEATURE_REQUIRED, diff_cu_qp_delta_depth should be the value of   (log2 CtbSizeY - 3 - log2_min_cu_qp_delta_size_minus3) , 

sorry about my previous comments, I did not have a deeper thinking about it.

Reply:
@XinfengZhang  My intent is that we should have the other attribute, but I'm trying to avoid getting bogged down on exactly what form it needs to take here when I don't have all of the information about what it needs to look like.  If `diff_cu_qp_delta_depth = 0` works with all existing hardware then I think it's good enough for these attributes, since that is the current state anyway.

To answer 1, that comment to always use zero would be changed once the new attribute is introduced to say that you should look at the new attribute, or use zero if it is not present.  This is only a problem if some future hardware is introduced which doesn't support zero at all, but that is broken anyway with the current definition of the API (where people have generally hardcoded Gen9 values).

For 2, based on the documentation Gen11 still supports zero, and if Gen12 is similar then it does too?  (Please correct me if this is wrong.)  So that works now, and the future attribute would then be able to offer the other possibilities as well (for the Gen11 example, the choice between 0 and 3 when using 64x64 CTUs).  Putting an Intel-only special case into this attribute when we are going to replace it soon after anyway does not seem like a good step to me.

Reply:
@XinfengZhang  No it doesn't.  If `cu_qp_delta_enabled_flag` is not set then `delta_qp()` never codes anything regardless of the level it happens at and the QP is fixed - see 7.3.8.14.  The inference you quote from 7.4.3.3.1 only exists because it is convenient in 8.6.1 to have the `Log2MinCuQpDeltaSize` for quantisation groups which isn't the whole slice.

The reason not to add partial fields now is that they would be deprecated when we add a proper attribute later, and I don't think it is a good idea to be maintaining an extra incomplete source of the same information.  If you want to include that information in this PR, what do you think of the suggestion above of a `VAConfigAttribValEncHEVCQPDeltaDepthSizes` attribute?

Reply:
yes, thanks, if cu_qp_delta_enabled_flag  == 1 and diff_cu_qp_delta_depth  == 0, it means there are only one syntax unit  delta_qp() for maximum TU when maximum TU is selected.  if cu_qp_delta_enabled_flag = 0  , no delta_qp() present.
 
add new VAConfigAttribValEncHEVCQPDeltaDepthSizes , there will be enough bits to describe this capability. but  as I mentioned, I am not very sure whether diff_cu_qp_delta_depth = 0 works on Gen12 VDEnc, even there are similar description with G11. need some test to verify it. @Tianhaol , please help to check.

about "but that is broken anyway with the current definition of the API (where people have generally hardcoded Gen9 values" , it is true for GST & FFmpeg.  because we have no such capability report, GST & FFmpeg may set it to 0 by default , but I suppose MSDK  set it to 3 by default for G12 VDEnc, if we add new attributes, MSDK also need to query and handle it. so it will be set to 0 instead.   suppose PSNR /SSIM will be impact , and impact validation cases.

about "proper attribute" , you mean that "min/max" is not enough for some cases, so after adding these, still need "proper attribute" then deprecate "min/max". what's the case? want to support different setting for different CTB size?

Reply:
If software written for one driver rather than against generic VAAPI is already using special knowledge about the behaviour of the driver it runs on then I don't see why it couldn't continue to do that.  (Correct me if I'm wrong, but I assume there is no intent that the Intel Media SDK would run with, say, the Mesa VAAPI driver in future.)

Min/max is not sufficient for the 0/3 case described by @Tianhaol and in the Ice Lake documentation, so if we want to support that we need an attribute with more fine-grained information.

Reply:
hi @fhvwy , @Tianhaol  has do some test, we try to use 0, 1, 2, 3 . it works. so maybe we could still keep previous design?log2_min_cu_qp_delta_size_minus3